[c10d] Make DebugInfoWriter Singleton across all PG objects

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #116489

Previously, we have the writer register to each NCCL PG(backend), so for every pg, we have a NCCL PG instance, so if we use some customized writer when multiple sub-PGs are used, we need to ensure user to register the writer for every backend which indicates a bad UX. Furthermore, the debug info is global, so it does not make sense to have the writer for each instance. We even have a static mutex in the `dumpDebuggingInfo` to ensure we serialize the write, that makes it more obvious that we can make the writer a singleton so that we only have one writer instance for all PG instances.

Although the rationale is clear, the implementation may vary a lot. So this PR is RFC for now to see if this implementation makes sense or not.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @wz337 @tianyu-l @wconstab @yf225

fduwjj committed:
[RFC][c10d] Make DebugInfoWriter Singleton across all PG objects

[ghstack-poisoned]

fduwjj committed:
Update on "[RFC][c10d] Make DebugInfoWriter Singleton across all PG objects"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fduwjj committed:
Update on "[RFC][c10d] Make DebugInfoWriter Singleton across all PG objects"


Previously, we have the writer register to each NCCL PG(backend), so for every pg, we have a NCCL PG instance, so if we use some customized writer when multiple sub-PGs are used, we need to ensure user to register the writer for every backend which indicates a bad UX. Furthermore, the debug info is global, so it does not make sense to have the writer for each instance. We even have a static mutex in the `dumpDebuggingInfo` to ensure we serialize the write, that makes it more obvious that we can make the writer a singleton so that we only have one writer instance for all PG instances.

Although the rationale is clear, the implementation may vary a lot. So this PR is RFC for now to see if this implementation makes sense or not.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fduwjj committed:
Update on "[RFC][c10d] Make DebugInfoWriter Singleton across all PG objects"


Previously, we have the writer register to each NCCL PG(backend), so for every pg, we have a NCCL PG instance, so if we use some customized writer when multiple sub-PGs are used, we need to ensure user to register the writer for every backend which indicates a bad UX. Furthermore, the debug info is global, so it does not make sense to have the writer for each instance. We even have a static mutex in the `dumpDebuggingInfo` to ensure we serialize the write, that makes it more obvious that we can make the writer a singleton so that we only have one writer instance for all PG instances.

Although the rationale is clear, the implementation may vary a lot. So this PR is RFC for now to see if this implementation makes sense or not.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

kwen2501 commented on torch/csrc/distributed/c10d/TraceUtils.h:
```diff
@@ -293,6 +287,29 @@ void DebugInfoWriter::write(const std::string& ncclTrace) {
+  if (writer_ == nullptr) {
+    std::string fileNamePrefix = getCvarString(
+        {"TORCH_NCCL_DEBUG_INFO_TEMP_FILE"}, "/tmp/nccl_trace_rank_");
+    std::unique_ptr<DebugInfoWriter> writerPtr(
+        new DebugInfoWriter(fileNamePrefix, rank));
```
Are we using `std::unique_ptr` here to auto-delete the writer object when the pointer itself is destroyed? If so, can we put down a comment here? Thanks!

kwen2501 reviewed (approved):
Overall looks good to me. Added some minor comments.

fduwjj commented on torch/csrc/distributed/c10d/NCCLUtils.hpp:
```diff
@@ -179,12 +179,20 @@ std::string getNcclErrorDetailStr(
   std::string filename_;
+
+ private:
+  static std::unique_ptr<DebugInfoWriter> writer_;
+  static bool hasWriterRegistered_;
```
Well.... I think about that. Actually we only allow register once, maybe call_once is a better choice?

wconstab replied:
i think technically, ke is right. if 2 threads called register, both of them could race through the code and both register/set the variables.

It would be uncommon for this to happen, since most of the time there is one python process per trainer and there would be GIL protecting multiple accesses from the python side anyway.  From the c++ side, there is only rank0's watchdog that calls dump API and implicitly triggers setting the default writer.  There could still be a race between c++ setting default writer and python setting custom writer.

fduwjj replied:
On c++ side, we have a static mutex to ensure the write is serialized, so I guess that should be fine? And for python, I agree, it's less likely that more than one thread is calling it. Well, user set customized writer in python before training, so there should not have the race. (cpp registration is lazy until timeout dump)

wconstab replied:
you could add a mutex. or you could use an atomic bool and use a CAS operation inside the register function, ensuring only the thread that changed the value of the bool is allowed to modify the writer

fduwjj replied:
OK, let me make it an atomic bool for now and we can always change it down the road.

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1209,13 +1200,8 @@ bool ProcessGroupNCCL::dumpDebuggingInfo() {
-      registerDebugInfoWriter(std::move(debugInfoWriterPtr));
-    }
-    debugInfoWriter_->write(ncclTrace);
+    DebugInfoWriter& writer = DebugInfoWriter::getWriter(globalRank());
```
we should document the API behavior in the py doc.

iiuc there are a couple of non-obvious rules that are in play
1) if you register once, you can't register again (e.g. to change writer)
2) if you don't register before something calls dumpDebuggingInfo, then you lost your chance to register (and the filesystem writer will be auto-registered).  This should be fine as you're expected to register at trainer startup, before any dumping would occur.
3) we don't offer 'unregister' because it doesn't seem likely anyone would have that use case.

I think these are ok (at least for now), just document them.

fduwjj replied:
Got it, sounds good to me.

wconstab commented on torch/csrc/distributed/c10d/NCCLUtils.hpp:
```diff
@@ -179,12 +179,20 @@ std::string getNcclErrorDetailStr(
-  DebugInfoWriter(int rank);
   virtual ~DebugInfoWriter();
   virtual void write(const std::string& ncclTrace);
+  static DebugInfoWriter& getWriter(int rank);
```
making these methods static is OK.  but it is not required, and could spoil using the `logPrefix()` function for example.  It does not hurt to leave these as instance methods, which can still access the static variables but also can access any instance variables.

fduwjj replied:
Actually no... This is actually expected because we don't allow calling constructor directly for a singleton. that's why both get and set are static.

fduwjj committed:
Update on "[c10d] Make DebugInfoWriter Singleton across all PG objects"


Previously, we have the writer register to each NCCL PG(backend), so for every pg, we have a NCCL PG instance, so if we use some customized writer when multiple sub-PGs are used, we need to ensure user to register the writer for every backend which indicates a bad UX. Furthermore, the debug info is global, so it does not make sense to have the writer for each instance. We even have a static mutex in the `dumpDebuggingInfo` to ensure we serialize the write, that makes it more obvious that we can make the writer a singleton so that we only have one writer instance for all PG instances.

Although the rationale is clear, the implementation may vary a lot. So this PR is RFC for now to see if this implementation makes sense or not.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge

kwen2501 commented on torch/csrc/distributed/c10d/TraceUtils.h:
```diff
@@ -293,6 +287,31 @@ void DebugInfoWriter::write(const std::string& ncclTrace) {
 }
 
+DebugInfoWriter& DebugInfoWriter::getWriter(int rank) {
+  if (writer_ == nullptr) {
```
we made `hasWriterRegistered_` atomic. Yet we are still checking `writer_` here.

fduwjj replied:
Well I can change this check to check the atomic variable but that does not fully protect the setWriter.
